multi-stark/src/verifier.rs

```diff
@@ -0,0 +1,226 @@
+use alloc::vec;
+use alloc::vec::Vec;
+
+use p3_air::Air;
+use p3_challenger::{CanObserve, FieldChallenger};
+use p3_commit::{Pcs, PolynomialSpace};
+use p3_field::{BasedVectorSpace, PrimeCharacteristicRing};
+use p3_uni_stark::{
+    SymbolicAirBuilder, VerificationError, VerifierConstraintFolder, get_log_quotient_degree,
+    recompose_quotient_from_chunks, verify_constraints,
+};
+use p3_util::log2_strict_usize;
+use p3_util::zip_eq::zip_eq;
+use tracing::instrument;
+
+use crate::config::{
+    Challenge, Domain, MultiStarkGenericConfig as MSGC, PcsError, Val, observe_base_as_ext,
+};
+use crate::proof::MultiProof;
+
+#[instrument(skip_all)]
+pub fn verify_multi<SC, A>(
+    config: &SC,
+    airs: &[A],
+    proof: &MultiProof<SC>,
+    public_values: &[Vec<Val<SC>>],
+) -> Result<(), VerificationError<PcsError<SC>>>
+where
+    SC: MSGC,
+    A: Air<SymbolicAirBuilder<Val<SC>>> + for<'a> Air<VerifierConstraintFolder<'a, SC>>,
+    Challenge<SC>: BasedVectorSpace<Val<SC>>,
+{
+    let MultiProof {
+        commitments,
+        opened_values,
+        opening_proof,
+        degree_bits,
+    } = proof;
+
+    let pcs = config.pcs();
+    let mut challenger = config.initialise_challenger();
+
+    // ZK mode is not supported yet
+    if config.is_zk() != 0 {
+        panic!("p3-multi-stark: ZK mode is not supported yet");
+    }
+
+    // Sanity checks
+    if airs.len() != opened_values.instances.len()
+        || airs.len() != public_values.len()
+        || airs.len() != degree_bits.len()
+    {
+        return Err(VerificationError::InvalidProofShape);
+    }
+
+    // Observe the number of instances up front to match the prover's transcript.
+    let n_instances = airs.len();
+    observe_base_as_ext::<SC>(&mut challenger, Val::<SC>::from_usize(n_instances));
+
+    // Validate opened values shape per instance (critical soundness check)
+    let log_quotient_degrees: Vec<usize> = airs
+        .iter()
+        .zip(public_values.iter())
+        .map(|(air, pv)| get_log_quotient_degree::<Val<SC>, A>(air, 0, pv.len(), config.is_zk()))
+        .collect();
+
+    for (i, air) in airs.iter().enumerate() {
+        let air_width = A::width(air);
+        let inst_vals = &opened_values.instances[i];
```

Comment:
Nit:
```suggestion
        let inst_opened_vals = &opened_values.instances[i];
```
so it's clear later in the code that these are the opened values